File: go/0707M-设计链表.go
package main

type Node struct {
	Val  int
	Next *Node
	Pre  *Node
}

// 循环双链表
type MyLinkedList struct {
	dummy *Node
}

// 创建虚拟头结点
func Constructor() MyLinkedList {
	rear := &Node{
		Val:  -1,
		Next: nil,
		Pre:  nil,
	}
	rear.Next = rear
	rear.Pre = rear
	return MyLinkedList{rear}
}

func (this *MyLinkedList) Get(index int) int {
	head := this.dummy.Next
	for head != this.dummy && index > 0 {
		head = head.Next
		index--
	}
	if index != 0 { // 循环一圈都没找到
		return -1
	}
	return head.Val
}

func (this *MyLinkedList) AddAtHead(val int) {
	dummy := this.dummy
	node := &Node{
		Val:  val,
		Next: dummy.Next,
		Pre:  dummy,
	}
	dummy.Next.Pre = node
	dummy.Next = node
}

func (this *MyLinkedList) AddAtTail(val int) {
	dummy := this.dummy
	node := &Node{
		Val:  val,
		Next: dummy,
		Pre:  dummy.Pre,
	}
	dummy.Pre.Next = node
	dummy.Pre = node
}

func (this *MyLinkedList) AddAtIndex(index int, val int) {
	head := this.dummy.Next
	for head != this.dummy && index > 0 {
		head = head.Next
		index--
	}
	node := &Node{
		Val:  val,
		Next: head,
		Pre:  head.Pre,
	}
	head.Pre.Next = node
	head.Pre = node
}

func (this *MyLinkedList) DeleteAtIndex(index int) {
	head := this.dummy.Next
	if head == this.dummy {
		return
	}
	for head.Next != this.dummy && index > 0 {
		head = head.Next
		index--
	}
	if index == 0 {
		head.Next.Pre = head.Pre
		head.Pre.Next = head.Next
	}
}

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Get(index);
 * obj.AddAtHead(val);
 * obj.AddAtTail(val);
 * obj.AddAtIndex(index,val);
 * obj.DeleteAtIndex(index);
 */

File: cpp/0707M-设计链表.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-13 09:56:40
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class MyLinkedList {
    struct ListNode {
        int val;
        ListNode *next;
        ListNode() : val(0), next(nullptr) {}
        ListNode(int x) : val(x), next(nullptr) {}
        ListNode(int x, ListNode *next) : val(x), next(next) {}
    };

   private:
    ListNode *dummy;
    int size;

   public:
    MyLinkedList() {
        dummy = new ListNode(0);
        size = 0;
    }

    int get(int index) {
        if (index < 0 || index >= size) return -1;
        ListNode *node = dummy->next;
        while (index--) {
            node = node->next;
        }
        return node->val;
    }

    void addAtHead(int val) { addAtIndex(0, val); }

    void addAtTail(int val) { addAtIndex(size, val); }

    void addAtIndex(int index, int val) {
        if (index > size) return;
        if (index < 0) index = 0;
        size++;
        ListNode *node = dummy;  // 前驱
        while (index--) node = node->next;
        ListNode *nNode = new ListNode(val);
        nNode->next = node->next;
        node->next = nNode;
    }

    void deleteAtIndex(int index) {
        if (index < 0 || index >= size) return;
        size--;
        ListNode *node = dummy;  // 前驱
        while (index--) node = node->next;
        node->next = node->next->next;
    }
};

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * MyLinkedList* obj = new MyLinkedList();
 * int param_1 = obj->get(index);
 * obj->addAtHead(val);
 * obj->addAtTail(val);
 * obj->addAtIndex(index,val);
 * obj->deleteAtIndex(index);
 */